### ISSUE TITLE
`n_fft` docstrings

### ISSUE THREAD

**Issue by lostanlen** [librosa/librosa#1522]
**Describe the bug**
`n_fft` is one of the most ubiquitous parameter in librosa, and yet its definition is relatively lapidary.

Here's a quick list i just made by grepping the source:
1. FFT window size for ``S``
2.  FFT window size
3. number of FFT bins to use, if ``y`` is provided.
4. length of the windowed signal after padding with zeros.
5. The number of samples per frame in D.
6. STFT window size
7.  If ``S`` is provided, then ``n_fft`` is inferred from ``S``. Else, copied from input 
8. number of FFT components in the resulting STFT
9. number of FFT components
10. number of FFT bins
11. The length of each analysis frame.
12. FFT window size for use in ``feature()`` if ``S`` is not provided.
(might contain more)

I would love it if someone could audit these definitions and see what can be improved. I'm not saying that we need the same definition everywhere (calling it a window size in some places and a number of bins somewhere else makes perfect sense) but i don't think we need as many as the ones i listed above.

Relatedly, it seems like some of these definitions are off by a factor two, since they refer to (1+n_fft//2) (assuming n_fft even) and not n_fft itself. The ISTFT even/odd disjunction (in the wake of #1244) is not very well explained.

**Comment by bmcfee** [comment]
Possibly related: #695

**Comment by bmcfee** [comment]
Tacking onto this, it might also be a good time to revise and update the documentation glossary.  Or remove it - it only got 40 views total across all versions in the last month.

**Comment by candidered** [comment]
Hello !
I start working on this issue, is it okay if I replace n_fft with long unambiguous terms like ''s_fft_window_size'' or ''fft_window_size_positive_offset_option" ?
Should I ask here every time I am unsure of which definition of n_fft I am looking at ? For instance I am quite lost finding out which definitions are used in constantq.py .

**Comment by lostanlen** [comment]
Hello @candidered, the issue is only about documentation. To answer your question, we are not planning to change the keyword `n_fft`, which is already well-established in Python and beyond. 

 > Should I ask here every time I am unsure of which definition of n_fft I am looking at ?

yes, if you're working on this issue, feel free to report on progress and ask questions.

> For instance I am quite lost finding out which definitions are used in constantq.py .

you'd have to look at the code, on a case-by-case basis. We can chat on the Scientific Python Discord if you wish

**Comment by bmcfee** [comment]
> is it okay if I replace n_fft with long unambiguous terms like ''s_fft_window_size'' or ''fft_window_size_positive_offset_option" ?

I don't think this is a good idea.  The API is stable at this point, and changing variable names in this way would A) not add much clarity, and B) break a ton of existing code.

Remember that the issue originally posted here is relating only to documentation, not API changes.

**Comment by NotNotoginseng** [comment]
Is this issue still open? I'd like to contribute.

**Comment by brookcs3** [comment]
I went through every `n_fft` parameter docstring on `main` to map the variants listed above: 36 definitions across 14 files. They reduce to three semantic groups.

**1. Analysis window length** (`stft`, the reassigned family, `fft_frequencies`, `get_duration`, `onset_strength_multi`, all of `feature/spectral`, `filters.mel`/`filters.chroma`, `piptrack`, `estimate_tuning`, the `feature.inverse` functions): here `n_fft` is the FFT window length in samples. It is currently worded six different ways ("FFT window size", "length of the FFT window", "STFT window size", "number of FFT components", "number of FFT bins", "number of FFT components in the resulting STFT"). The bins/components wordings are the factor-of-two confusion this issue calls out, since `n_fft=2048` gives `1 + n_fft // 2 = 1025` bins.

**2. Samples per frame of a spectrogram input** (`istft`, `phase_vocoder`, `griffinlim`, `specshow`): inferred as `2 * (shape[-2] - 1)`, with the odd-window caveat. `specshow` already states the formula plainly; `istft`'s explanation is the one quoted in this issue.

**3. Time-offset parameter** (`frames_to_time` and the five siblings in `core/convert`): "length of the FFT window... offset of `n_fft // 2`". These six are already identical and clear.

I have a PR ready that:
- standardizes group 1 on "length of the FFT window" (the wording `mfcc` and `melspectrogram` already use, consistent with `stft`'s fuller definition);
- rewords the bins/components cases and states the `1 + n_fft // 2` relationship where a filter bank or output is actually shaped by it (`filters.mel`, `filters.chroma`, `feature.inverse.*`);
- rewrites `istft`'s inference note with the explicit formula;
- fixes one factor-of-two error found along the way: `mel_to_stft` documents its return as shape `(..., n_fft, t)`, but `nnls(mel_basis, M)` produces `(..., 1 + n_fft // 2, t)`;
- leaves group 3 untouched.

Happy to adjust any of the canonical wordings.

**Comment by bmcfee** [comment]
I think this is now sufficiently resolved, so i'll close this out.
